src/Libraries/Microsoft.Extensions.AI.Evaluation/CompositeEvaluator.cs

```diff
@@ -0,0 +1,175 @@
+            {
+                if (!metricNames.Add(metricName))
+                {
+                    Throw.ArgumentException(nameof(evaluators), $"Cannot add multiple evaluators for '{metricName}'.");
+                }
+            }
+        }
+
+        EvaluationMetricNames = metricNames;
+        _evaluators = [.. evaluators];
+    }
+
+    /// <summary>
+    /// Evaluates the supplied <paramref name="modelResponse"/> and returns an <see cref="EvaluationResult"/>
+    /// containing one or more <see cref="EvaluationMetric"/>s.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The <see cref="EvaluationMetric.Name"/>s of the <see cref="EvaluationMetric"/>s contained in the returned
+    /// <see cref="EvaluationResult"/> should match <see cref="EvaluationMetricNames"/>.
+    /// </para>
+    /// <para>
+    /// Also note that <paramref name="chatConfiguration"/> must not be omitted if one or more composed
+    /// <see cref="IEvaluator"/>s use an AI model to perform evaluation.
+    /// </para>
+    /// </remarks>
+    /// <param name="messages">
+    /// The conversation history including the request that produced the supplied <paramref name="modelResponse"/>.
+    /// </param>
+    /// <param name="modelResponse">The response that is to be evaluated.</param>
+    /// <param name="chatConfiguration">
+    /// A <see cref="ChatConfiguration"/> that specifies the <see cref="IChatClient"/> and the
+    /// <see cref="IEvaluationTokenCounter"/> that should be used if one or more composed <see cref="IEvaluator"/>s use
+    /// an AI model to perform evaluation.
+    /// </param>
+    /// <param name="additionalContext">
+    /// Additional contextual information (beyond that which is available in <paramref name="messages"/>) that composed
+    /// <see cref="IEvaluator"/>s may need to accurately evaluate the supplied <paramref name="modelResponse"/>.
+    /// </param>
+    /// <param name="cancellationToken">
+    /// A <see cref="CancellationToken"/> that can cancel the evaluation operation.
+    /// </param>
+    /// <returns>An <see cref="EvaluationResult"/> containing one or more <see cref="EvaluationMetric"/>s.</returns>
+    public async ValueTask<EvaluationResult> EvaluateAsync(
+        IEnumerable<ChatMessage> messages,
+        ChatMessage modelResponse,
+        ChatConfiguration? chatConfiguration = null,
+        IEnumerable<EvaluationContext>? additionalContext = null,
+        CancellationToken cancellationToken = default)
+    {
+        var metrics = new List<EvaluationMetric>();
+
+        IAsyncEnumerable<EvaluationResult> resultsStream =
+            EvaluateAndStreamResultsAsync(
+                messages,
+                modelResponse,
+                chatConfiguration,
+                additionalContext,
+                cancellationToken);
+
+        await foreach (EvaluationResult result in resultsStream.ConfigureAwait(false))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I haven't been commenting on all of these ThrowIfCancellationRequested, but it's very unusual to see them used so heavily. Consider searching and deleting any that aren't truly necessary.

Reply:
Thanks for catching this - I've audited and removed all redundant calls to `ThrowIfCancellartionRequested`.